lib/dpl/provider/pages.rb

```diff
@@ -44,34 +53,140 @@ def slug
+        context.shell "git init" or raise 'Could not create new git repo'
+        print_step 'Repo created successfully'
+        context.shell "git checkout --orphan '#{@target_branch}'" or raise 'Could not create new git repo'
+        print_step "An orphan branch #{@target_branch} created successfully"
+      end
+
+      def github_configure
+        print_step "Configuring git committer to be #{@gh_name} <#{@gh_email}>"
         context.shell "git config user.email '#{@gh_email}'"
         context.shell "git config user.name '#{@gh_name}'"
+      end
+
+      def github_commit
+        print_step "Preparing to deploy #{@target_branch} branch to gh-pages"
+        context.shell "touch \"deployed at `date` by #{@gh_name}\""
         context.shell "echo '#{@gh_fqdn}' > CNAME" if @gh_fqdn
-        context.shell 'git add .'
-        context.shell "FILES=\"`git commit -m 'Deploy #{@project_name} to #{@gh_ref}:#{@target_branch}' | tail`\"; echo \"$FILES\"; echo \"$FILES\" | [ `wc -l` -lt 10 ] || echo '...'"
-        context.shell "git push --force --quiet 'https://#{@gh_token}@#{@gh_ref}' master:#{@target_branch} > /dev/null 2>&1"
+        context.shell 'git add -A .'
+        context.shell "git commit#{@git_commit_opts} -qm 'Deploy #{@project_name} to #{@gh_ref}:#{@target_branch}'"
+        context.shell 'git show --stat-count=10 HEAD'
+      end
+
+      def github_deploy
+        print_step "Doing the git push..."
+        unless context.shell "git push#{@git_push_opts} --quiet '#{@gh_remote_url}' '#{@target_branch}':'#{@target_branch}' &>/dev/null"
+          error "Couldn't push the build to #{@gh_ref}:#{@target_branch}"
+        end
+        print_step "It should've been pushed by now"
+        return
+      end
+
+      def prepare_dir_tree(dir)
+            build = "#{dir}/build"
+            work = "#{dir}/work"
+
+            Dir.mkdir(build)
+            print_step "Created a temporary build directory #{build}"
+
+            Dir.mkdir(work)
+            print_step "Created a temporary working directory #{work}"
+
+            return build, work
+      end
+
+      def prepare_build_dir(build, tmp)
+            FileUtils.cp_r("#{build}/.", tmp)
+            print_step "Copied over build contents to #{tmp}"
+            FileUtils.cd(tmp, :verbose => @verbose) do
+                FileUtils.rm_r '.git', :force => true, :verbose => @verbose # cleanup garbage
+                print_step "Cleaned up .git artifacts"
+            end
       end
 
       def push_app
-        Dir.mktmpdir {|tmpdir|
-            FileUtils.cp_r("#{@build_dir}/.", tmpdir)
-            FileUtils.cd(tmpdir, :verbose => true) do
-              unless github_deploy
-                error "Couldn't push the build to #{@gh_ref}:#{@target_branch}"
+        print_step "Starting deployment of #{@target_branch} branch to gh-pages..."
+        print_step "The deployment is configured to preserve the target branch if it exists on remote"
```

Comment:
This needs to be conditional on `@keep_history` if you keep that parameter.

Reply:
oops, I probably lost it during rebase